### PR TITLE
Improve `active_plugins` filter

### PR DESCRIPTION
**Pull request by swissspidy**
See #518 and #597.

Filters active plugins to only include required ones. This means:

* The plugin being tested
* All dependencies of the plugin being tested
* Plugin Check itself
* All active dependents of Plugin Check (they could be adding new checks)

To-do:

* [x] Add Behat tests (ideally after #518 is merged).
	* IMO such end-to-end tests are best for testing this scenario.

### CODE REVIEW

**Comment by ernilambar** [comment]
Currently `Requires at least:` is set to `6.3`. We should set to `6.5` as `WP_Plugin_Dependencies` was introduced in the 6.5 release.

**Comment by swissspidy** [comment]
Thanks for pointing that out. Alternatively, I could also just add some `class_exists()` checks.

Also, we should really be running PHP unit tests against our minimum required WP version...

Edit: #610

**Review by ernilambar — APPROVED** [review]

File: includes/Checker/Preparations/Force_Single_Plugin_Preparation.php [hunk 1]
```diff
@@ -74,39 +75,67 @@ public function prepare() {
...
+		if ( ! in_array( $this->plugin_basename, $active_plugins, true ) ) {
+			return $active_plugins;
+		}
 
-			$plugin_base_file = plugin_basename( $plugin_check_file );
+		if ( defined( 'WP_PLUGIN_CHECK_MAIN_FILE' ) ) {
+			$plugin_check_file = WP_PLUGIN_CHECK_MAIN_FILE;
+		} else {
+			$plugins_dir       = defined( 'WP_PLUGIN_DIR' ) ? WP_PLUGIN_DIR : WP_CONTENT_DIR . '/plugins';
+			$plugin_check_file = $plugins_dir . '/plugin-check/plugin.php';
```
Comment (felixarntz) [inline includes/Checker/Preparations/Force_Single_Plugin_Preparation.php]: Not strictly related to this PR since the code was already here before, but: Rather than doing this guesswork, wouldn't it be more reliable to look from here?
```suggestion
			$plugin_check_file = dirname( __DIR__, 3 ) . '/plugin.php';
```

Looks ugly, but should definitely be more reliable.

File: includes/Checker/Preparations/Force_Single_Plugin_Preparation.php [hunk 2]
```diff
@@ -74,39 +75,67 @@ public function prepare() {
...
+			);
+
+			$new_active_plugins = array_merge(
+				$new_active_plugins,
+				// Include any dependents of Plugin Check, but only if they were already active.
+				array_filter(
+					WP_Plugin_Dependencies::get_dependents( dirname( $plugin_check_basename ) ),
+					static function ( $dependent ) use ( $active_plugins, $new_active_plugins ) {
+						return in_array( $dependent, $active_plugins, true ) ||
+								in_array( $dependent, $new_active_plugins, true );
```
Comment (felixarntz) [inline includes/Checker/Preparations/Force_Single_Plugin_Preparation.php]: Why do we need the second condition here? That seems pointless to me, since we're checking for existence in `$new_active_plugins` which is the same array we want to potentially add these plugins to.
```suggestion
						return in_array( $dependent, $active_plugins, true );
```

File: includes/Checker/Preparations/Force_Single_Plugin_Preparation.php [hunk 3]
```diff
@@ -74,39 +75,67 @@ public function prepare() {
...
+			$new_active_plugins = array_merge(
+				$new_active_plugins,
+				WP_Plugin_Dependencies::get_dependencies( $this->plugin_basename )
+			);
+
+			$new_active_plugins = array_merge(
+				$new_active_plugins,
+				// Include any dependents of Plugin Check, but only if they were already active.
+				array_filter(
+					WP_Plugin_Dependencies::get_dependents( dirname( $plugin_check_basename ) ),
```
Comment (felixarntz) [inline includes/Checker/Preparations/Force_Single_Plugin_Preparation.php:129]: What a confusing API, requiring a plugin _file_ for `get_dependencies()` but a plugin _slug_ for `get_dependents()`... 🙄

File: includes/Checker/Preparations/Force_Single_Plugin_Preparation.php [hunk 4]
```diff
@@ -74,39 +75,67 @@ public function prepare() {
...
+	 * Filters active plugins to only include required ones.
+	 *
+	 * This means:
+	 *
+	 * * The plugin being tested
+	 * * All dependencies of the plugin being tested
+	 * * Plugin Check itself
+	 * * All plugins depending on Plugin Check (they could be adding new checks)
 	 *
 	 * @since 1.0.0
```
Comment (felixarntz) [inline includes/Checker/Preparations/Force_Single_Plugin_Preparation.php:87]: Can you add a `@since 1.2.0|n.e.x.t` here to cover this enhancement?

**Review by felixarntz — APPROVED** [review]
@swissspidy Thanks, looks great! Just a few minor comments.

**Comment by felixarntz** [comment]
@swissspidy Are you planning to merge #518 first and then see which uncommented Behat tests this here will fix?

File: includes/Checker/Preparations/Force_Single_Plugin_Preparation.php (hunk 1 above)
Comment (swissspidy) [reply includes/Checker/Preparations/Force_Single_Plugin_Preparation.php]: Good point. Downside is that this needs to be kept in mind in case files are moved.

File: includes/Checker/Preparations/Force_Single_Plugin_Preparation.php (hunk 2 above)
Comment (swissspidy) [reply includes/Checker/Preparations/Force_Single_Plugin_Preparation.php]: The dependent could be the plugin being checked, so it would already be in `$new_active_plugins`.

But I'm also doing a `array_unique()` below, so it shouldn't matter.

**Comment by swissspidy** [comment]
> @swissspidy Are you planning to merge #518 first and then see which uncommented Behat tests this here will fix?

Yep!

File: includes/Checker/Preparations/Force_Single_Plugin_Preparation.php (hunk 2 above)
Comment (felixarntz) [reply includes/Checker/Preparations/Force_Single_Plugin_Preparation.php]: Yeah, that's what I was thinking. I was initially hesitant about using `array_unique()` (because before this was just one single string comparison instead), but since you use it, no need to prevent duplicates here either.

**Comment by swissspidy** [comment]
Right now I can't enable any of the commented out Behat tests because they're still not passing, because we still need to address (2) and (3) from https://github.com/WordPress/plugin-check/issues/597#issuecomment-2327483452

**Comment by felixarntz** [comment]
> Right now I can't enable any of the commented out Behat tests because they're still not passing, because we still need to address (2) and (3) from [#597 (comment)](https://github.com/WordPress/plugin-check/issues/597#issuecomment-2327483452)

I think it would only be point (3) that would help with the remaining Behat tests, so maybe we can work on that one first.

Could you give it a try to implement a Behat test that runs two runtime checks, one from the addon and one built-in one? For instance amend/duplicate the very last Behat test in `plugin-check.feature` that way. I'm curious what happens then, since part of the problem that (3) would solve is only relevant if the only runtime check comes from an addon. But if we also have a runtime check from the built-in ones, that's a different implication because it'll already know we need to initialize runtime check relevant logic.

**Comment by felixarntz** [comment]
Following up on my previous comment, based on what I wrote in https://github.com/WordPress/plugin-check/issues/597#issuecomment-2332289958, I would actually assume that if you do what I recommended before, you'll run into the exception that the slug is invalid - which will prevent the rest from proceeding even if then the plugin checker technically would know to initialize runtime environment preparations.

**Comment by swissspidy** [comment]
> > Right now I can't enable any of the commented out Behat tests because they're still not passing, because we still need to address (2) and (3) from [#597 (comment)](https://github.com/WordPress/plugin-check/issues/597#issuecomment-2327483452)
> 
> I think it would only be point (3) that would help with the remaining Behat tests, so maybe we can work on that one first.
> 
> Could you give it a try to implement a Behat test that runs two runtime checks, one from the addon and one built-in one? For instance amend/duplicate the very last Behat test in `plugin-check.feature` that way. I'm curious what happens then, since part of the problem that (3) would solve is only relevant if the only runtime check comes from an addon. But if we also have a runtime check from the built-in ones, that's a different implication because it'll already know we need to initialize runtime check relevant logic.

Argh, sorry I missed this suggestion. I think we can do this in #612 though
